cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```diff
@@ -0,0 +1,275 @@
+            Message message = iterator.next();
+            lastFetchedMessageId = message.getId();
+            try {
+              Notification notification = GSON.fromJson(new String(message.getPayload(), StandardCharsets.UTF_8),
+                                                        Notification.class);
+              fetchedNotifications.add(notification);
+            } catch (JsonSyntaxException e) {
+              LOG.warn("Failed to decode message with id {}. Skipped. ", message.getId(), e);
+              // lastFetchedMessageId was updated at the beginning of the loop to skip this message in next fetch
+            }
+          }
+        }
+
+        // Sleep for configured number of milliseconds if there are no notifications
+        if (lastFetchedMessageId == null) {
+          return cConf.getLong(Constants.Scheduler.EVENT_POLL_DELAY_MILLIS);
+        }
+
+        final String lastMessageId = lastFetchedMessageId;
+        // Execute processed notifications in one transaction
+        Retries.supplyWithRetries(new Supplier<Void>() {
+          @Override
+          public Void get() {
+            Transactions.executeUnchecked(transactional, new TxCallable<Void>() {
+              @Override
+              public Void call(DatasetContext context) throws Exception {
+                processFetchedNotifications(context, fetchedNotifications);
+                updateMessageId(context, lastMessageId);
+                return null;
+              }
+            });
+            return null;
+          }
+        }, co.cask.cdap.common.service.RetryStrategies.fixDelay(Constants.Retry.RUN_RECORD_UPDATE_RETRY_DELAY_SECS,
+                                                                TimeUnit.SECONDS));
+
+        // Transaction was successful, so reset the failure count to 0
+        failureCount = 0;
+        // The last fetched message id was persisted successfully, update the local field as well.
+        messageId = lastMessageId;
+        return 0L;
+
+      } catch (ServiceUnavailableException e) {
+        SAMPLING_LOG.warn("Failed to contact service {}. Will retry in next run.", e.getServiceName(), e);
+      } catch (TopicNotFoundException e) {
+        SAMPLING_LOG.warn("Failed to fetch from TMS. Will retry in next run.", e);
+      } catch (Exception e) {
+        LOG.warn("Failed to get and process notifications. Will retry in next run", e);
+      }
+
+      // If there is any failure during fetching of notifications or looking up of schedules,
+      // delay the next fetch based on the strategy
+      // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+      return retryStrategy.nextRetry(++failureCount, 0);
+    }
+
+    /**
+     * Processes a list of notifications
+     *
+     * @param context the dataset context
+     * @param notifications the list of pairs of message ids and fetched notifications from TMS
+     * @throws Exception
```

Comment:
nit: list IOException and DatasetManagementException individually and state when will they be thrown